### PR TITLE
infra: use sval to parse javascript instead of mdx's `evaluate`

### PR DESCRIPTION
**Pull request by ovflowd**
This PR is an extension of the Next.js 15 Pull Request with an experimental approach of using `sval` to evaluate dynamic JavaScript expressions during runtime instead of the regular Browser's `eval`.

### CODE REVIEW

**Comment by AugustinMauroy** [comment]
for better maintainability put these code on `lib/mdx` so `lib/mdx/compile.mjs` and `lib/mdx/evaluate.mjs` _recommended by V0_

Also there are no reason to put it on root instead of putting it on dedicated place

**Comment by avivkeller** [comment]
On Vercel, it's throwing a `500` error. Locally, I got the same error, with the following details:
```
@node-core/website:dev:  ⨯ Error: JSXElement isn't implemented
@node-core/website:dev:     at evaluate (/apps/site/.next/server/chunks/ssr/08b5e_sval_dist_sval_7f0539.js:1610:19)
@node-core/website:dev:     at /apps/site/.next/server/chunks/ssr/08b5e_sval_dist_sval_7f0539.js:6886:26
@node-core/website:dev:     at AsyncLocalStorage.run (node:async_hooks:346:14)
```

I'm assuming it's referring to the `type` of `expression`: `JSXElement` (in some cases)

In `recma`, it looks like https://github.com/mdx-js/recma/tree/main/packages/recma-build-jsx is used under-the-hood (maybe?).

**Comment by avivkeller** [comment]
IIUC Recma is a transformer, whereas SVal is an interpreter. I don't think these are interoperable.

**Comment by ovflowd** [comment]
> On Vercel, it's throwing a `500` error. Locally, I got the same error, with the following details:
> 
> ```
> @node-core/website:dev:  ⨯ Error: JSXElement isn't implemented
> @node-core/website:dev:     at evaluate (/apps/site/.next/server/chunks/ssr/08b5e_sval_dist_sval_7f0539.js:1610:19)
> @node-core/website:dev:     at /apps/site/.next/server/chunks/ssr/08b5e_sval_dist_sval_7f0539.js:6886:26
> @node-core/website:dev:     at AsyncLocalStorage.run (node:async_hooks:346:14)
> ```
> 
> I'm assuming it's referring to the `type` of `expression`: `JSXElement` (in some cases)
> 
> In `recma`, it looks like [mdx-js/recma@`main`/packages/recma-build-jsx](https://github.com/mdx-js/recma/tree/main/packages/recma-build-jsx?rgh-link-date=2024-11-15T22%3A09%3A38Z) is used under-the-hood (maybe?).

This PR is a draft :)

> In recma, it looks like [mdx-js/recma@main/packages/recma-build-jsx](https://github.com/mdx-js/recma/tree/main/packages/recma-build-jsx?rgh-link-date=2024-11-15T22%3A09%3A38Z) is used under-the-hood (maybe?).

Afaik Recma uses `eval` -- but I might be wrong. The error of `Error: JSXElement isn't implemented` refers to the fact that `sval` does not support these `estree` node types. I opened an issue here: https://github.com/Siubaak/sval/issues/107. However, I doubt that `sval` should handle interpreting them and should do the same that the recma plugin does to bind them to `__jsx` probably.

**Comment by ovflowd** [comment]
The issue comes from these lines: https://github.com/nodejs/nodejs.org/blob/main/apps/site/pages/en/index.mdx?plain=1#L19-L39

Where `rehype-react`'s `mdast-util-mdx-expression`/`hast-util-to-jsx-runtime` will transform these into Expressions and ask for `sval` to evaluate them, which is fine, but then it reaches to these node types that `mdast-util-mdx-expression` [...]

The issue is that these are unraveled within `sval` if not, I could bind these nodes directly to `hast-util-to-jsx-runtime`, but of course `sval` attempts to evaluate them within its own `estree` pipeline.

So the maintainer of `sval` will need to make a patch there... If you find another way of bypassing that and being able to intercept before it attempts to evaluate those pieces... LMK.

**Comment by ovflowd** [comment]
Note that ironically our home pages are (afaik) the only MDX pages we have that use Expression statements, and technically _we could_ change our MDX sources to simply not do that;

But then we are pretty much hiding ourselves on a partial Evaluater implementation and any time that we need to have MDX JavaScript expressions that have React Fragments or Elements inside, it would break.

So, I'm open to suggestions here 🤔

**Comment by avivkeller** [comment]
> Afaik Recma uses eval -- but I might be wrong

I don't think so. Looking at the Recma source code, it's really doing the following:
1. `esast-util-from-js` (Which is basically a wrapper around `acorn`, which doesn't rely on `eval`)
2. `estree-util-build-jsx` (Which doesn't handle any raw JS syntax, and doesn't use eval) [ref](https://github.com/syntax-tree/estree-util-build-jsx/blob/main/lib/index.js)
3. `estree-util-to-js` (Which uses `astring`, which doesn't appear to use `eval`) [ref](https://github.com/davidbonnet/astring/blob/main/src/astring.js)

**Comment by ovflowd** [comment]
> > Afaik Recma uses eval -- but I might be wrong
> 
> 
> 
> I don't think so. Looking at the Recma source code, it's really doing the following:
> 
> 1. `esast-util-from-js` (Which is basically a wrapper around `acorn`, which doesn't rely on `eval`)
> 
> 2. `estree-util-build-jsx` (Which doesn't handle any raw JS syntax, and doesn't use eval) [ref](https://github.com/syntax-tree/estree-util-build-jsx/blob/main/lib/index.js)
> 
> 3. `estree-util-to-js` (Which uses `astring`, which doesn't appear to use `eval`) [ref](https://github.com/davidbonnet/astring/blob/main/src/astring.js)

Gotcha, I thought it was inside recma, but @wooorm told me on a DM that mdjx-js/mdx uses eval in some part of its pipeline. Hence they whole plan of going with "sval" -- which to be honest it is in its infancy. (I just don't recall where)

Also I don't think that specific library you mentioned before has eval 🤔

**Comment by avivkeller** [comment]
Okay, I looked into it a bit more. The part of `mdx-js` that uses eval is likely <https://github.com/mdx-js/mdx/blob/main/packages/mdx/lib/run.js>. We do use this in this file, see line 31. Looks like we don't even use Recma for any evaluation.

**Comment by ovflowd** [comment]
> Okay, I looked into it a bit more. The part of `mdx-js` that uses eval is likely [mdx-js/mdx@`main`/packages/mdx/lib/run.js](https://github.com/mdx-js/mdx/blob/main/packages/mdx/lib/run.js?rgh-link-date=2024-11-15T22%3A51%3A02Z). We do use this in this file, see line 31. Looks like we don't even use Recma for any evaluation.

Oh yes, it calls `new Function()`

**Comment by avivkeller** [comment]
We could use Recma (no eval) to process the source code and JSX, and then pass that into a safe evaluator, like SVal

**Comment by ovflowd** [comment]
> s the source code and JSX, and then pass that into a safe

The issue is that the Expressions are not evaluated yet, so at that point, we have `ExpressionStatement`, whose `retype-react` passes to slab, as it wants a broken-down tree (updated node tree) with the expressions evaluated... Inside those expressions is where it unravels the components, before that it doesn't know about their existence.

I could be wrong, so please feel free to play around with this PR.

**Comment by avivkeller** [comment]
Okay, I did some fiddling around, and I got it evaluating without the browser `eval`, however the output isn't correct yet:
image


And "Hydration failed because the server rendered HTML didn't match the client. As a result this tree will be regenerated on the client. This can happen if a SSR-ed Client Component used"

**Comment by avivkeller** [comment]
got it working! Mind if I push directly to your PR?

**Comment by ovflowd** [comment]
> got it working! Mind if I push directly to your PR?

Sure!

**Comment by avivkeller** [comment]
:-/ I wonder why Vercel is failing? I'm able to build fine locally

**Review by avivkeller — APPROVED** [review]

File: apps/site/package.json [hunk 1]
```diff
@@ -79,6 +79,9 @@
     "postcss-simple-vars": "~7.0.1",
     "react": "^18.3.1",
     "react-dom": "^18.3.1",
+    "recma": "^1.0.0",
```
Comment (ovflowd) [inline apps/site/package.json]: Are these deps needed?

Comment (avivkeller) [reply apps/site/package.json]: Oh, your right. Sorry, I added them when I was messing around with Recma, I'll push a commit removing them

**Comment by ovflowd** [comment]
We did it!

**Pull request merged**